src/librustc/ty/mod.rs

```diff
@@ -758,62 +758,121 @@ impl ty::EarlyBoundRegion {
     }
 }
 
+#[derive(Clone, Debug, RustcEncodable, RustcDecodable)]
+pub enum GenericParamDef {
+    Lifetime(RegionParamDef),
+    Type(TypeParamDef),
+}
+
+impl GenericParamDef {
+    pub fn index(&self) -> u32 {
+        match self {
+            GenericParamDef::Lifetime(lt) => lt.index,
+            GenericParamDef::Type(ty)     => ty.index,
+        }
+    }
+
+    pub fn get_type(&self) -> Option<TypeParamDef> {
```

Comment:
I would prefer it if this wasn't a thing, and everyone was pattern-matching. I'm not even sure about the prevalence of `if let` because that won't help `const` generics.